pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+			// (Should never happen.)
+			log.Warningf("InsertIdleSummaryAllocations: failed to find %s: %s", prop, key)
+			return
+		}
+
+		idleTotals[key].CPUCost -= sa.CPUCost
+		idleTotals[key].GPUCost -= sa.GPUCost
+		idleTotals[key].RAMCost -= sa.RAMCost
+	})
+
+	// Turn remaining idle totals into idle allocations and insert them.
+	for key, rt := range idleTotals {
+		idleAlloc := &SummaryAllocation{
+			Name: fmt.Sprintf("%s/%s", key, IdleSuffix),
+			Properties: &AllocationProperties{
+				Cluster: rt.Cluster,
+				Node:    rt.Node,
+			},
+			Start:   rt.Start,
+			End:     rt.End,
+			CPUCost: rt.CPUCost,
+			GPUCost: rt.GPUCost,
+			RAMCost: rt.RAMCost,
+		}
+
+		sas.Insert(idleAlloc)
+	}
+
+	return nil
+}
+
+// Delete removes the allocation with the given name from the set
+func (sas *SummaryAllocationSet) Delete(name string) {
+	if sas == nil {
+		return
+	}
+
+	sas.Lock()
+	defer sas.Unlock()
+
+	delete(sas.externalKeys, name)
+	delete(sas.idleKeys, name)
+	delete(sas.SummaryAllocations, name)
+}
+
+// Each invokes the given function for each SummaryAllocation in the set
+func (sas *SummaryAllocationSet) Each(f func(string, *SummaryAllocation)) {
+	if sas == nil {
+		return
+	}
+
+	for k, a := range sas.SummaryAllocations {
+		f(k, a)
+	}
+}
+
+// IdleAllocations returns a map of the idle allocations in the AllocationSet.
+func (sas *SummaryAllocationSet) IdleAllocations() map[string]*SummaryAllocation {
+	idles := map[string]*SummaryAllocation{}
+
+	if sas == nil || len(sas.SummaryAllocations) == 0 {
+		return idles
+	}
+
+	sas.RLock()
+	defer sas.RUnlock()
+
+	for key := range sas.idleKeys {
+		if sa, ok := sas.SummaryAllocations[key]; ok {
+			idles[key] = sa // TODO Clone()?
+		}
+	}
+
+	return idles
+}
+
+// Insert aggregates the current entry in the SummaryAllocationSet by the given Allocation,
+// but only if the Allocation is valid, i.e. matches the SummaryAllocationSet's window. If
+// there is no existing entry, one is created. Nil error response indicates success.
+func (sas *SummaryAllocationSet) Insert(sa *SummaryAllocation) error {
+	if sas == nil {
+		return fmt.Errorf("cannot insert into nil SummaryAllocationSet")
+	}
+
+	sas.Lock()
+	defer sas.Unlock()
+
+	if sas.SummaryAllocations == nil {
+		sas.SummaryAllocations = map[string]*SummaryAllocation{}
+	}
+
+	if sas.externalKeys == nil {
+		sas.externalKeys = map[string]bool{}
+	}
+
+	if sas.idleKeys == nil {
+		sas.idleKeys = map[string]bool{}
+	}
+
+	// Add the given Allocation to the existing entry, if there is one;
+	// otherwise just set directly into allocations
+	if _, ok := sas.SummaryAllocations[sa.Name]; ok {
+		sas.SummaryAllocations[sa.Name].Add(sa)
+	} else {
+		sas.SummaryAllocations[sa.Name] = sa
+	}
+
+	// If the given Allocation is an external one, record that
+	if sa.IsExternal() {
+		sas.externalKeys[sa.Name] = true
+	}
+
+	// If the given Allocation is an idle one, record that
+	if sa.IsIdle() {
+		sas.idleKeys[sa.Name] = true
+	}
```

Comment:
Pathological case: non-external/idle allocation gets inserted with name `k`, then another allocation with name `k` is inserted that is marked as external/idle. What should happen?

Reply:
Well, we intentionally try our best to avoid that by appending suffices onto those (`__external__` and `__idle__`). So, we'd ultimately end up adding the two together and calling it all idle. IMO that's fine. Please, people, don't name your containers "__idle__" ahah.